### DISCUSSION TITLE
1.16.5 - Nightly: pistons, equalization, displacing, downhill flowing

### DISCUSSION THREAD

**Discussion by Felicis** [Felicis/Water-Physics-Overhaul#28] [Nightly Releases]
THIS VERSION IS NIGHTLY: backup your worlds, expect crashes! Do not post issues for this, instead join the discussion :)

Hi y'all,
Finally there are few bugs that I feel like I can show you this version. Skds-core has been integrated into this mod and is therefore no longer needed.
There are still many bugs and there are still some disabled/unimplemented features, but have a look at what is already working (and maybe you even find bugs I have not found yet).

Here is a short summary of this version (tested only with water):

- pistons and placing/falling blocks displaces fluids (displace distance configurable)
  - known bug: pistons can still destroy fluids when the fluids can not flow away
- horizontal equalization (configurable distance)
  - known bug: when a pit is full fluids start dancing above it
- fluid flowing downhill (slide distance configurable)
- buckets can take and place fluid (bucket reach distance configurable)
- fluid loggable blocks accept fluid (not just waterloggable blocks, but most non-full blocks: I will upload the list somewhere later)
  - can hold fluid
  - fluid can flow through it
  - accept water when placed/falling or pushed by pistons

disabled:
- fluid mixing (lava + water)
- no water placing in nether

untested/probably buggy:
- lava, forge fluids
- fluid ticking is not fixed yet: if a fluid does not equalize/flow downhill, you can force a tick by placing and breaking a block next to fluid
- mod compat

<hr /><em>This discussion was created from the release <a href='https://github.com/Felicis/Water-Physics-Overhaul/releases/tag/0.1.3-alpha'>1.16.5 - Nightly: pistons, equalization, displacing, downhill flowing</a>.</em>

**Comment by djmrFunnyMan** [comment]
Instantly crashes with exit code 1
![ezgif-5-8f63fe8b6f](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/49e851b5-b256-40d5-ba07-eaab5bd2d988)

**Reply by Felicis** [reply]
I can't read the error message in the video, but try adding the mod [Statement (v4.2.5 forge)](https://www.curseforge.com/minecraft/mc-mods/statement/files/4275870). It isn't used in the code any more, but it might be needed to start it.

**Reply by djmrFunnyMan** [reply]
Yup that was it. It starts now

**Reply by Felicis** [reply]
nice, that was my bad. I'll remove it until the next nightly.

**Comment by djmrFunnyMan** [comment]
Oh wow. Well right off the bat (unrelated to fluids)
1- Placing blocks has no animation
2- Destroying blocks in survival doesn't drop anything
3- You have infinite blocks in survival

**Reply by Felicis** [reply]
I think I know the problem there....

**Reply by djmrFunnyMan** [reply]
4- Extreme lag spikes when placing down water. (only if theres no lower ground for it to flow for some reason)

https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/98825a27-0c44-43c8-94c5-38b84094a9e0


5- Water plays the flowing sound endlessly
6- Odd physics (updating this block did not fix this)
![2023-06-30_18 29 37](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/14bf8821-8c8e-4f0b-9ff4-1a070f0931f8)
7- Extreme lag in normal world (compared to the previous version of the mod)
![2023-06-30_18 40 36](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/37e43373-249b-4dee-92f5-fd1496f26c55)
8- This weird thing with kelp
![2023-06-30_18 38 45](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/01ef1bda-e2f2-4f6b-b5b2-2e1708976159)
9- Waterlogged ravines spawn with vegetation but no water?
![2023-06-30_18 42 50](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/974fc303-7141-42b0-9e0b-c415bab00819)

**Comment by Felicis** [comment]
> 4- Extreme lag spikes when placing down water. (only if theres no lower ground for it to flow for some reason)
> 7- Extreme lag in normal world (compared to the previous version of the mod)

Haven't looked at performance yet. Remind me to check how it fares when _equalization_ and _fluid ticking_ is fixed...

> 5- Water plays the flowing sound endlessly

sounds are not correctly implemented yet. I'll let you know when to check for it

> 6- Odd physics (updating this block did not fix this)

this looks like the "dancing" water equalization I mentioned. I will fix the equalization for the next nightly and let's see if it stops dancing... 

> 8- This weird thing with kelp

kelp (and also seagrass) is on the bug list :+1: 

> 9- Waterlogged ravines spawn with vegetation but no water?

I have yet to debug and fix worldgen. remind me once I publish a nightly with _worldgen fixes_

**Reply by djmrFunnyMan** [reply]
> kelp (and also seagrass) is on the bug list 👍

The weird thing is on the version I played before (1.18) kelp and seagrass never had any issues. I assume it's some problem with 1.16 then

**Reply by djmrFunnyMan** [reply]
> Haven't looked at performance yet.

One thing. I was still playing on my old config with high equalization and sliding (64 and 32). After switching to default settings issue 4 disapeared and 7 got better.
I suppose it is expected that increasing these decreases performance but the performance loss wasn't this drastic before this version.

**Reply by Felicis** [reply]
Nah, it's just that I had to almost rewrite the mod from scratch. And the fluid system is completely different: 

In the old version there were a) fluid blocks (with different fluid levels) and b) waterlogged blocks (always water source). Effectively only fluid blocks contained _real, flowing fluids_, whereas waterlogged blocks had all kinds of bugs and problems.

In the new system _every_ block has a (possibly empty) fluid, which is stored in the chunk and savegame just like blocks. The tricky part is that the fluid and the block at each coordinate have to be kept _in sync_. 
You can check that they are in sync by checking in F3 under _targeted fluid_ that the fluid and its block match: either a) a fluid with its fluid block (like in the screenshot) or b) a fluid in a waterlogged block. 
![image](https://github.com/Felicis/Water-Physics-Overhaul/assets/11877488/bdfe1caa-f8a3-4a72-8a53-a6f6bd574869)

**Reply by Felicis** [reply]
oh wow, yeah those are high settings :sweat_smile: we'll have to see in future if they can't be sped up

**Reply by djmrFunnyMan** [reply]
> oh wow, yeah those are high settings

Technically not even. The config says can both can go as high as 256 💀

**Comment by djmrFunnyMan** [comment]
About sliding distance rewrite:
1 - Water now doesn't "wake up" when a hole appears. Previously if you've made a hole near some water it would start sliding towards it. Now you need to update the water itself before it realizes there's a hole to slide towards. Random ticks cannot make it slide either
![2023-07-01_12 54 07](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/bd28fe53-a7c4-45a0-9ceb-1a496ee9f876)

2 - I cannot test much further because the /fill command is broken with fluids? 
![2023-07-01_13 04 33](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/0c591821-d2c0-46b3-8b79-9e80333972e9)
However I can see that it's much better when it comes to issue #21 
![2023-07-01_13 11 46](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/6edc1381-e07d-49d8-9daf-16189f06a531)

**Comment by djmrFunnyMan** [comment]
There is also a big issue of client/server desynch. Creating ghost fluids, duplicating fluids etc (reloading a world corrects those)
1- Example of client "duping" water (it's not actually real on server side)

https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/d6698c54-b9d4-485d-91e1-5583ecfa5f68

2- Examples of ghost blocks when placing fluids with buckets (this is what issue 6- Odd physics actually was) these blocks disappear when reloading the world
![2023-07-01_12 57 23](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/06511aae-519e-4f5f-ab3d-5cbc4c3b253a)
![2023-07-01_13 12 04](https://github.com/Felicis/Water-Physics-Overhaul/assets/68905776/c46fc70c-deca-4754-abac-a4d875ee35b5)
